File: app/src/main/java/com/centaury/cataloguemovie/ui/tvshow/TVShowFragment.java
package com.centaury.cataloguemovie.ui.tvshow;


import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentActivity;
import androidx.lifecycle.ViewModelProvider;
import androidx.lifecycle.ViewModelProviders;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.centaury.cataloguemovie.R;
import com.centaury.cataloguemovie.di.Injectable;
import com.centaury.cataloguemovie.utils.CommonUtils;
import com.facebook.shimmer.ShimmerFrameLayout;

import java.util.Locale;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.Unbinder;

/**
 * A simple {@link Fragment} subclass.
 */
public class TVShowFragment extends Fragment implements Injectable {

    @BindView(R.id.rv_tv_show)
    RecyclerView mRvTvshow;
    @BindView(R.id.shimmer_view_container)
    ShimmerFrameLayout mShimmerViewContainer;
    @BindView(R.id.txt_load_more)
    TextView mTxtLoadMore;
    @Inject
    ViewModelProvider.Factory factory;
    private Unbinder unbinder;

    public TVShowFragment() {
        // Required empty public constructor
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        View view = inflater.inflate(R.layout.fragment_tvshow, container, false);

        unbinder = ButterKnife.bind(this, view);
        return view;
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        if (getActivity() != null) {
            TVShowViewModel tvShowViewModel = obtainViewModel(getActivity());
            String language = Locale.getDefault().toLanguageTag();

            TVShowAdapter tvShowAdapter = new TVShowAdapter(getActivity());
            tvShowViewModel.getTVShows().observe(this, tvShowResultsItems -> {
                tvShowAdapter.submitList(tvShowResultsItems);
                tvShowAdapter.notifyDataSetChanged();
            });
            tvShowViewModel.getGenreTVShow(language).observe(this, genresItemList -> {
                tvShowAdapter.setListGenreTVShow(genresItemList);
                tvShowAdapter.notifyDataSetChanged();
            });

            tvShowViewModel.getLoadingState().observe(this, loadingState -> {
                if (loadingState) {
                    mShimmerViewContainer.startShimmer();
                    mShimmerViewContainer.setVisibility(View.VISIBLE);
                } else {
                    mShimmerViewContainer.stopShimmer();
                    mShimmerViewContainer.setVisibility(View.GONE);
                }
            });

            tvShowViewModel.getLoadMoreLoadingState().observe(this, loadMore -> {
                if (loadMore) {
                    mTxtLoadMore.setVisibility(View.VISIBLE);
                } else {
                    mTxtLoadMore.setVisibility(View.GONE);
                }
            });

            mRvTvshow.setLayoutManager(new LinearLayoutManager(getContext()));
            mRvTvshow.setHasFixedSize(true);
            mRvTvshow.setAdapter(tvShowAdapter);
            mRvTvshow.addItemDecoration(new CommonUtils.TopItemDecoration(55));
        }
    }

    @NonNull
    private TVShowViewModel obtainViewModel(FragmentActivity activity) {
        return ViewModelProviders.of(activity, factory).get(TVShowViewModel.class);
    }

    @Override
    public void onResume() {
        super.onResume();
        mShimmerViewContainer.startShimmer();
    }

    @Override
    public void onPause() {
        mShimmerViewContainer.stopShimmer();
        super.onPause();
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        unbinder.unbind();
    }
}

File: app/src/main/java/com/centaury/cataloguemovie/ui/tvshow/TVShowFragment.kt
package com.centaury.cataloguemovie.ui.tvshow

import android.annotation.SuppressLint
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.LinearLayoutManager
import com.centaury.cataloguemovie.MovieCatalogueApp
import com.centaury.cataloguemovie.databinding.FragmentTvshowBinding
import com.centaury.cataloguemovie.di.component.DaggerDiscoverTVShowComponent
import com.centaury.cataloguemovie.ui.detail.DetailMovieActivity
import com.centaury.cataloguemovie.ui.main.ItemClickCallback
import com.centaury.cataloguemovie.utils.CommonUtils
import com.centaury.cataloguemovie.utils.LoaderState
import com.centaury.cataloguemovie.utils.showToast
import com.centaury.cataloguemovie.utils.timberE
import com.centaury.domain.model.Genre
import com.centaury.domain.model.TVShow
import javax.inject.Inject

/**
 * A simple [Fragment] subclass.
 */
class TVShowFragment : Fragment(), ItemClickCallback {

    @Inject
    lateinit var viewModelFactory: ViewModelProvider.Factory

    private val tvShowViewModel: TVShowViewModel by viewModels {
        viewModelFactory
    }

    private lateinit var binding: FragmentTvshowBinding

    private var tvShowData = arrayListOf<TVShow>()
    private var genreData = arrayListOf<Genre>()
    private val tvShowAdapter: TVShowAdapter by lazy {
        TVShowAdapter(tvShowData, genreData, this)
    }

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        // Inflate the layout for this fragment
        binding = FragmentTvshowBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        initView(binding)
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        initInjector()
    }

    private fun initInjector() {
        DaggerDiscoverTVShowComponent.builder()
            .appComponent((requireActivity().application as MovieCatalogueApp).appComponent)
            .build()
            .inject(this)
    }

    private fun initView(binding: FragmentTvshowBinding) {
        with(binding.rvTvShow) {
            hasFixedSize()
            layoutManager = LinearLayoutManager(context)
            addItemDecoration(CommonUtils.TopItemDecoration(55))
            adapter = tvShowAdapter
        }

        initObserver(binding)
    }

    @SuppressLint("NotifyDataSetChanged")
    private fun initObserver(binding: FragmentTvshowBinding) {
        tvShowViewModel.state.observe(viewLifecycleOwner) { state ->
            when (state) {
                is LoaderState.ShowLoading -> {
                    binding.shimmerViewContainer.startShimmer()
                    binding.hasTVShows = true
                }
                is LoaderState.HideLoading -> {
                    binding.shimmerViewContainer.stopShimmer()
                    binding.hasTVShows = false
                }
            }
        }

        tvShowViewModel.result.observe(viewLifecycleOwner) { result ->
            tvShowData.clear()
            tvShowData.addAll(result)
            CommonUtils.toggleEmptyState(result.size, binding.emptyState, binding.rvTvShow)
            tvShowAdapter.notifyDataSetChanged()
        }

        tvShowViewModel.error.observe(viewLifecycleOwner) { error ->
            timberE(error)
            context?.showToast(error)
        }

        tvShowViewModel.resultGenre.observe(viewLifecycleOwner) { resultGenre ->
            genreData.clear()
            genreData.addAll(resultGenre)
            tvShowAdapter.notifyDataSetChanged()
        }

        tvShowViewModel.errorGenre.observe(viewLifecycleOwner) { errorGenre ->
            timberE(errorGenre)
            context?.showToast(errorGenre)
        }

    }

    override fun onResume() {
        super.onResume()
        binding.shimmerViewContainer.startShimmer()
    }

    override fun onPause() {
        binding.shimmerViewContainer.stopShimmer()
        super.onPause()
    }

    private fun showTransitionImage(movieId: Int, image: ImageView, title: String) {
        val intent = Intent(activity, DetailMovieActivity::class.java).apply {
            putExtra(DetailMovieActivity.DETAIL_EXTRA_TV_SHOW, movieId)
        }

        activity?.let {
            it.startActivity(
                intent,
                CommonUtils.pairOptionsTransition(it, image, title).toBundle()
            )
        }
    }

    override fun onItemClick(movieId: Int, image: ImageView, title: String) {
        showTransitionImage(movieId, image, title)
    }
}